modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/ImageMosaicSourceElement.java

```diff
@@ -0,0 +1,233 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2020, Open Source Geospatial Foundation (OSGeo)
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.gce.imagemosaic;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.util.logging.Level;
+import java.util.regex.Pattern;
+import org.apache.commons.io.FilenameUtils;
+import org.geotools.gce.imagemosaic.catalogbuilder.CatalogBuilderConfiguration;
+import org.geotools.gce.imagemosaic.properties.PropertiesCollector;
+
+/** An elementary source element (a File, a URL, ...) of an ImageMosaic */
+abstract class ImageMosaicSourceElement<T> {
+
+    /** Return the inner object */
+    abstract T getInnerObject();
```

Comment:
Maybe a better name?

```suggestion
    abstract T getElement();
```